src/composite_core.jl

```diff
@@ -0,0 +1,124 @@
+struct PrimalAdditionFailedException{P} <: Exception
+    primal::P
+    differential::Composite{P}
+    original::Exception
+end
+
+function Base.showerror(io::IO, err::PrimalAdditionFailedException{P}) where {P}
+    println(io, "Could not construct $P after addition")
+    print("This probably means no default constructor (")
+    print(P, "(", join(propertynames(err.differential), ", "), ")")
+    println(") is defined.\n Either define a default constructor, or")
+    print("overload `ChainRulesCore.construct(::Type{$P}, ")
+    println(" ::$(typeof(err.differential)))`,")
+    println("or overload `Base.:+(::$P, ::$(typeof(err.differential)))`.")
+    println("Original Exception:")
+    print(io, err.original)
+    println()
+end
+
+
+
+"""
+    backing(x)
+
+Returns a version of `x` that is of the type the backing of a `Composite` is.
+i.e. as a `NamedTuple` or `Tuple`.
+
+This is kind of the opposite of `construct` for structs.
+"""
+backing(x::Tuple) = x
+backing(x::NamedTuple) = x
+backing(x::Composite) = getfield(x, :backing)
+
+function backing(x::T)::NamedTuple where T
+    nfields = fieldcount(T)
+    names = ntuple(ii->fieldname(T, ii), nfields)
+    types = ntuple(ii->fieldtype(T, ii), nfields)
+
+    if @generated
+        vals = ntuple(ii->:(getfield(x, $ii)), nfields)
+        return :(NamedTuple{$names, Tuple{$(types...)}}($vals))
+    else
+        vals = ntuple(ii->getfield(x, ii), nfields)
+        return NamedTuple{names, Tuple{types...}}(vals)
+    end
+end
+
+
+
+"""
+    construct(::Type{T}, fields::[NamedTuple|Tuple])
+
+Constructs an object of type `T`, with the given fields.
+Fields must be correct in name and type, and `T` must have a default constructor
+"""
+function construct(::Type{T}, fields::NamedTuple{L}) where {T, L}
+    # Tested and varified that that this avoids a ton of allocations
```

Comment:
do we have a test for this? if not is it because "low allocations" is too vague to test for?

Reply:
Good idea, we should add a test.